### ISSUE TITLE
[reward] android_gui: `extract_answer` fallback takes the first `0`/`1`/`2` anywhere in the reply, so a reply that is not a bare digit can be scored on the wrong digit

### ISSUE THREAD

**Issue by shaurya416** [hiyouga/EasyR1#11014]
`examples/reward_function/android_gui.py:46-49` (used by `compute_score` at `:73-80`), commit `602a128` (current main)

```python
    # 情况2: 响应包含多余文字，提取第一个出现的0/1/2
    match = re.search(r"[012]", response)
    if match:
        return match.group(0)
```

If the stripped reply is not exactly `"0"`, `"1"` or `"2"`, the fallback uses the first character in the text that matches `[012]`. The pattern has no digit boundary and no anchor. Any `0`/`1`/`2` that appears before the actual choice becomes the answer, for example a digit inside a card value (`12`, `2024`), a step number (`1.`) or a hedge (`0, 1, or 2`).

Scope: `android_gui.jinja` asks for a single digit and `max_response_length` is 64, so only replies that break that format reach the fallback. Bare-digit replies score correctly. I have not measured how often real rollouts break the format.

## Measured

The real `examples/reward_function/android_gui.py` from `602a128` was loaded with `importlib`. It has no heavy dependencies, so nothing was stubbed, installed or modified. A second harness copied `extract_answer` and `compute_score` verbatim into plain Python, and both gave the same results. The gold labels in `yuehua-s/numbergame` are strings (the `answer` column is typed string on the HF datasets-server), so the comparison is `str == str`. In the first 100 train rows the split is `'0'` 30, `'1'` 35, `'2'` 35.

| # | response | ground_truth | reward | expected |
|---|----------|--------------|--------|----------|
| c1 | `'1'` | `'1'` | 1.0 | 1.0 |
| c2 | `'0'` | `'1'` | 0.0 | 0.0 |
| c3 | `"I don't know"` | `'1'` | 0.0 | 0.0 |
| c4 | `''` | `'1'` | 0.0 | 0.0 |
| c5 | `'The answer is 1'` | `'1'` | 1.0 | 1.0 |
| c6 | `'I choose option 2'` | `'2'` | 1.0 | 1.0 |
| 7 | `'The numbers are 12, 45, 3. Green light so the largest, 45, is at position 2.'` | `'2'` | **0.0** | 1.0 |
| 8 | same reply | `'1'` | **1.0** | 0.0 |
| 9 | `'1. Green light. 2. Numbers 7, 3, 9. 3. Largest is 9. Output: 2'` | `'2'` | **0.0** | 1.0 |
| 10 | same reply | `'1'` | **1.0** | 0.0 |
| 11 | `'2024'` | `'2'` | **1.0** | 0.0 |
| 12 | `'10'` | `'1'` | **1.0** | 0.0 |
| 13 | `'0, 1, or 2'` | `'0'` | **1.0** | 0.0 |
| 14 | `'I refuse. Option 1 is unsafe'` | `'1'` | **1.0** | 0.0 |

- c1-c6 are controls: a bare digit, a wrong digit, no digit, an empty reply, and the two prose cases from the `__main__` block. All six behave as intended.
- In rows 7-10 the same reply scores 0.0 or 1.0 depending only on the label.
- `skip_special_tokens` defaults to True, so end-of-turn tokens do not reach the function. `'1<|im_end|>'` also scores correctly.

Modeled distribution (synthetic, not observed rollouts): replies were generated from the template `The cards show a, b and c. The light is <color>, so the answer is position P.`, with a, b and c distinct integers drawn from an assumed range of 1..99, over 20,000 draws. The mean reward was 0.444 when P is the correct position and 0.273 when P is a wrong position. The ideal values are 1.000 and 0.000. A second run with a different seed gave 0.449 and 0.274. These numbers show how the parser behaves on multi-number prose. They do not estimate what happens in real training.

## Consequence

The file docstring and `examples/android_gui_cookbook/README.md` state the scoring rule as correct +1.0, wrong 0.0. That rule holds only for bare-digit replies. When a reply breaks the format and contains another `0`/`1`/`2` before the chosen digit, the reward follows that earlier digit and not the choice. `verl/trainer/main.py:69-70` builds `reward_fn` and `val_reward_fn` from the same config, and `examples/qwen2_5_vl_3b_android_gui_grpo.sh` uses this function for both (rollout temperature 0.9). So for those replies, both the training reward and the reported validation accuracy stop tracking correctness. How much this affects real runs depends on how often rollouts break the format, and that has not been measured.

No test covers this path. `tests/` has no reward tests, and the `__main__` block only prints bare-digit and single-digit-sentence cases.

## Suggested fix

Take the last boundary-delimited match, as other code in this directory does:
- `dapo.py:125-126` takes the last match of an anchored pattern (`match[-1] if match else "[INVALID]"`).
- `math.py` grades through mathruler's `extract_boxed_content`, which uses `rfind` (the last `\boxed`).

Minimal change:

```python
matches = re.findall(r"(?<!\d)[012](?!\d)", response)
return matches[-1] if matches else ""
```

With the same harness, this leaves c1-c6 unchanged and fixes rows 7-10. `'2024'` and `'10'` (rows 11-12) extract `""` and score 0.0. Rows 13-14, a hedge or a refusal that mentions an option, stay ambiguous under any digit scan. The prompt requires a single digit, so `re.fullmatch(r"\s*([012])\s*", response)` returning `""` otherwise is a stricter option that follows that output contract and also rejects hedges.

I found no existing issue or PR about this. #568 added the file and has no review comments on it.

Happy to open the PR.
